aten/src/ATen/native/sparse/SoftMax.cpp

```diff
@@ -0,0 +1,647 @@
+    }
+    return;
+  }
+
+  auto nnz = values.size(0);
+  auto nvalues = get_nvalues(sizes, sparse_dim);
+
+  auto values_2 = values.view({nnz, nvalues});
+  auto values_accessor = values_2.accessor<scalar_t, 2>();
+
+  auto out_values_2 = out_values.view({out_nnz, nvalues});
+  auto out_values_accessor = out_values_2.accessor<scalar_t, 2>();
+
+  auto grad_values_2 = grad_values.view({grad_nnz, nvalues});
+  auto grad_values_accessor = grad_values_2.accessor<scalar_t, 2>();
+
+  /* Compute independent pools of indices */
+  auto pools = get_pools(out_indices, sizes, dim);
+
+  int64_t grain_size = 1;
+  parallel_for(0, pools.size(), grain_size, [&](int64_t begin, int64_t end) {
+      for (auto p = begin; p < end; p++) {
+        auto pool_indices = pools[p];
+
+        std::vector<scalar_t> tmp_row(nvalues, 0);
+
+        /* Compute tmp = - sum_j output_j * grad_j */
+        for (index_t i : pool_indices) {
+          auto out_values_row = out_values_accessor[i];
+          auto values_row = values_accessor[i];
+          auto low = std::lower_bound(grad_offsets.begin(), grad_offsets.end(), out_offsets[i]);
+          auto j = low - grad_offsets.begin();
+
+          if (j < grad_nnz && (out_offsets[i] == grad_offsets[j])) {
+            auto grad_values_row = grad_values_accessor[j];
+            for (int64_t k=0; k<nvalues; k++) {
+              if (LogSoftMax) {
+                tmp_row[k] -= grad_values_row[k];
+              } else {
+                tmp_row[k] -= out_values_row[k] * grad_values_row[k];
+              }
+            }
+          }
+        }
+
+        /* Compute grad_input = output * (grad + tmp)*/
+        for (index_t i : pool_indices) {
+          auto out_values_row = out_values_accessor[i];
+          auto values_row = values_accessor[i];
+          auto low = std::lower_bound(grad_offsets.begin(), grad_offsets.end(), out_offsets[i]);
+          auto j = low - grad_offsets.begin();
+
+          if (j < grad_nnz && (out_offsets[i] == grad_offsets[j])) {
+            auto grad_values_row = grad_values_accessor[j];
+            for (int64_t k=0; k<nvalues; k++) {
+              if (LogSoftMax) {
+                values_row[k] = grad_values_row[k] + std::exp(out_values_row[k]) * tmp_row[k];
+              } else {
+                values_row[k] = out_values_row[k] * (grad_values_row[k] + tmp_row[k]);
+              }
+            }
+          } else {
+            for (int64_t k=0; k<nvalues; k++) {
+              if (LogSoftMax) {
+                values_row[k] = std::exp(out_values_row[k]) * tmp_row[k];
+              } else {
+                values_row[k] = out_values_row[k] * (tmp_row[k]);
+              }
+            }
+          }
+        }
+      }
+    });
+}
+
+} // namespace
+
+Tensor softmax_sparse_cpu(const Tensor& input_, const int64_t dim_, const bool half_to_float) {
+  AT_ASSERT(input_.is_sparse());
+  AT_ASSERTM(!half_to_float, "softmax with half to float conversion is not supported on CPU");
```

Comment:
Building on the above, this one should be a `TORCH_CHECK`, since a user could reasonably run into this bug (right?)

Reply:
Fixed.